### PR TITLE
Simple short-form Self-RAG Pack

### PR DESCRIPTION
**Pull request by MarouaneMaatouk**
# Description

Simple short-form Self-RAG implementation using llama_cpp adapted from the author's [code](https://github.com/AkariAsai/self-rag)
Paper: https://arxiv.org/abs/2310.11511
Model: https://huggingface.co/m4r1/selfrag_llama2_7b-GGUF

Fixes #8502 (https://github.com/run-llama/llama_index/issues/8502)

## Type of Change

Please delete options that are not relevant.

- [x] New Loader/Tool
- [ ] Bug fix / Smaller change
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have added a library.json file if a new loader/tool was added
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by MarouaneMaatouk** [comment]
@anoopshrma could you please review this?

**Comment by anoopshrma** [comment]
> @anoopshrma could you please review this?

Hey! Sure will do by tomorrow. Currently away from machine.

File: llama_hub/llama_packs/self_rag/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Simple self-RAG short form pack
+
+This LlamaPack implements short form the [self-RAG paper by Akari et al.](https://arxiv.org/pdf/2310.11511.pdf).
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/README.md:3]: nit: implements (*in short form)

File: llama_hub/llama_packs/self_rag/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+Novel framework called Self-Reflective Retrieval-Augmented Generation (SELF-RAG). Which aims to enhance the quality and factuality of large language models (LLMs) by combining retrieval and self-reflection mechanisms.
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/README.md:5]: nit: This paper presents a novel framework called ...

File: llama_hub/llama_packs/self_rag/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+```
+
+From here, you can use the pack. You can import the relevant modules from the download folder (in the example below we assume it's a relative import or the directory 
+has been added to your system path).
+
+```python
+from self_rag_pack.base import SelfRAGQueryEngine
+
+query_engine = SelfRAGQueryEngine(model_path=model_path, retriever=retreiver, verbose=True)
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/README.md:46]: nit: *retriever

File: llama_hub/llama_packs/self_rag/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+response = query_engine.query("Who won best Director in the 1972 Academy Awards?")
+```
+
+You can also use/initialize the pack directly.
+
+```python
+from llm_compiler_agent_pack.base import SelfRAGPack
+
+agent_pack = SelfRAGPack(model_path=model_path, retriever=retreiver, verbose=True)
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/README.md:56]: nit: retriever

File: llama_hub/llama_packs/self_rag/base.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            ]
+        )
+    return isUse_score
+
+
+class SelfRAGQueryEngine(CustomQueryEngine):
+    """Simple short form self RAG query engine."""
+
+    llm: Any = Field(default=None, description="llm")
+    retreiver: BaseRetriever = Field(default=None, description="Retreiver")
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/base.py:161]: nit: retriever

File: llama_hub/llama_packs/self_rag/base.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                paragraphs_final_score, key=paragraphs_final_score.get
+            )
+            answer = llm_response_text[best_paragraph_id]
+            if self.verbose:
+                print_text(f"Selected the best answer: {answer}\n", color="blue")
+
+        answer = _postprocess_answer(answer)
+        if self.verbose:
+            print_text(f"Final answer: {answer}\n", color="green")
+        return Response(response=str(answer))
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/base.py:259]: is there a way to add source_nodes too? given the nature of the self-rag algo, would be awesome to track the final retrieved nodes/paragraphs

File: llama_hub/llama_packs/self_rag/base.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    ) -> None:
+        """Init params."""
+        super().__init__(verbose=verbose, **kwargs)
+        model_kwargs = model_kwargs or _MODEL_KWARGS
+        self.generate_kwargs = generate_kwargs or _GENERATE_KWARGS
+        try:
+            from llama_cpp import Llama  # noqa: F401
+        except ImportError:
+            raise ImportError(_IMPORT_ERROR_MSG)
+        self.llm = Llama(model_path=model_path, verbose=verbose, **model_kwargs)
```
Comment (jerryjliu) [inline llama_hub/llama_packs/self_rag/base.py:182]: we do have a llamacpp wrapper in llamaindex. out of curiosity is there a reason you are using the llama_cpp lib directly? (it's fine either way, just curious)

Comment (jerryjliu) [reply llama_hub/llama_packs/self_rag/base.py:182]: oh nvm seems like its for logprobs

**Review by jerryjliu — APPROVED** [review]
this is fantastic! some comments but afterwards excited to land and promote this

File: llama_hub/llama_packs/self_rag/base.py (hunk 6 above)
Comment (MarouaneMaatouk) [reply llama_hub/llama_packs/self_rag/base.py]: Added in the last commit

**Review by anoopshrma — APPROVED** [review]
Awesome @MarouaneMaatouk!

**Comment by anoopshrma** [comment]
Hey @jerryjliu , since all the comments have been resolved. Can I go ahead and merge it or do you want to take another look.

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**